src/com/facebook/buck/jvm/java/JarShape.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2013-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.facebook.buck.jvm.java;
+
+import com.facebook.buck.graph.AbstractBreadthFirstTraversal;
+import com.facebook.buck.rules.BuildRule;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.ImmutableSortedSet;
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public enum JarShape {
+  MAVEN {
+    @Override
+    public Summary gatherDeps(BuildRule root) {
+      // Our life is a lot easier if the root is a JavaLibrary :)
+      if (!(root instanceof JavaLibrary)) {
+        throw new RuntimeException("Boom");
+      }
+
+      // We only need the first order maven deps, since maven's depenedency resolution process will
+      // pull in any transitive deps. Gather candidate maven deps using a BFS
+      Set<HasMavenCoordinates> mavenDeps = new HashSet<>();
+      Set<JavaLibrary> combinedTransitiveDeps = new HashSet<>();
+      new AbstractBreadthFirstTraversal<JavaLibrary>((JavaLibrary) root) {
+        @Override
+        public Iterable<JavaLibrary> visit(JavaLibrary javaLibrary) throws RuntimeException {
+          if (!javaLibrary.equals(root) && javaLibrary.getMavenCoords().isPresent()) {
+            mavenDeps.add(javaLibrary);
+            // We may have a transitive dep on another maven dep.
+            combinedTransitiveDeps.addAll(
+                javaLibrary.getTransitiveClasspathDeps().stream()
```

Comment:
this, combined with the traversal, is O(n^2). How hard would it be to make it more efficient?